File: app/src/main/java/com/example/starwars/activity/FilmActivity.java
package com.example.starwars.activity;

import static com.example.starwars.activity.FilmListActivity.ERROR_MESSAGE;

import android.os.Bundle;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import com.bumptech.glide.Glide;
import com.example.starwars.R;
import com.example.starwars.StarWarsApp;
import com.example.starwars.model.Film;
import com.example.starwars.model.Genre;
import com.example.starwars.viewmodel.FilmActivityViewModel;

public class FilmActivity extends AppCompatActivity {

    private static final String IMAGE_BASE_URL = "https://image.tmdb.org/t/p/w342/";

    private FilmActivityViewModel mViewModel;

    private ImageView mFilmIcon;
    private TextView mFilmTitleTextView;
    private TextView mFilmDateTextView;
    private TextView mFilmGenresTextView;
    private TextView mFilmOverviewTextView;
    private int mId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.Theme_StarWars);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_film);

        mFilmIcon = findViewById(R.id.film_icon);
        mFilmTitleTextView = findViewById(R.id.film_title);
        mFilmDateTextView = findViewById(R.id.film_date);
        mFilmGenresTextView = findViewById(R.id.film_genres);
        mFilmOverviewTextView = findViewById(R.id.film_overview);

        mId = getIntent().getIntExtra(FilmListActivity.FILM_ID_EXTRA_KEY, 0);
        mViewModel = ViewModelProviders.of(this).get(FilmActivityViewModel.class);
        mViewModel.init(mId);
        mViewModel.getFilmLiveData().observe(this, new Observer<Film>() {
            @Override
            public void onChanged(Film film) {
                setData(film);
            }
        });

    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    private void setData(Film film) {
        if (film != null) {
            Glide.with(StarWarsApp.getContext())
                    .load(IMAGE_BASE_URL + film.getPosterPath())
                    .error(R.drawable.ic_launcher_background)
                    .into(mFilmIcon);
            mFilmTitleTextView.setText(film.getTitle());
            mFilmDateTextView.setText(film.getReleaseDate());
            mFilmGenresTextView.setText(Genre.convertToString(film.getGenres()));
            mFilmOverviewTextView.setText(film.getOverview());
        } else {
            Toast.makeText(this, ERROR_MESSAGE, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/starwars/activity/FilmActivity.kt
package com.example.starwars.activity

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.example.starwars.API_ENDPOINTS.IMAGE_BASE_URL
import com.example.starwars.ConnectionUtils
import com.example.starwars.R
import com.example.starwars.activity.FilmListActivity.Companion.TAG
import com.example.starwars.adapter.CrewAdapter
import com.example.starwars.databinding.ActivityFilmBinding
import com.example.starwars.model.Film
import com.example.starwars.model.Genre
import com.example.starwars.viewmodel.FilmActivityViewModel
import dagger.hilt.android.AndroidEntryPoint
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
class FilmActivity : AppCompatActivity() {

    private val mViewModel: FilmActivityViewModel by viewModels<FilmActivityViewModel>()
    @Inject lateinit var mAdapter: CrewAdapter
    private val binding by lazy { ActivityFilmBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        val id = intent.getIntExtra(FilmListActivity.FILM_ID_EXTRA_KEY, 0)
        mViewModel.init(id)
        mViewModel.loadData(ConnectionUtils.isInternetConnected(this))
        mViewModel.filmLiveData.observe(this,
            { t -> setData(t) })

        binding.filmCrewRv.adapter = mAdapter

        binding.filmToFullCrew.setOnClickListener {
            val intent = Intent(this@FilmActivity, CrewActivity::class.java)
            intent.putExtra(FilmListActivity.FILM_ID_EXTRA_KEY, id)
            startActivity(intent)
        }
    }

    private fun setData(film: Film?) {
        if (film != null) {

            if (!ConnectionUtils.isInternetConnected(this)) {
                Toast.makeText(this, FilmListActivity.CACHE_MESSAGE, Toast.LENGTH_SHORT).show()
            }
            Log.d(TAG, "setData: " + film.title)

            with(binding) {
                Glide.with(applicationContext)
                    .load(IMAGE_BASE_URL + film.posterPath)
                    .error(R.drawable.ic_launcher_background)
                    .into(filmIcon)
                filmTitle.text = film.title
                filmDate.text = film.releaseDate
                filmGenres.text = Genre.convertToString(film.genres)
                filmOverview.text = film.overview
                mAdapter.setData(film.crews)

            }
        } else {
            mAdapter.setData(null)
            Toast.makeText(this, FilmListActivity.ERROR_MESSAGE, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onResume() {
        super.onResume()

    }
}